source/include/dqm4hep/Directory.h

```diff
@@ -192,38 +220,38 @@ namespace dqm4hep {
     //-------------------------------------------------------------------------------------------------
 
     template <typename T>
-    inline Directory<T> *Directory<T>::parent() const
+    inline typename Directory<T>::DirectoryPtr Directory<T>::parent() const
     {
-      return m_pParent;
+      return m_parent;
     }
 
     //-------------------------------------------------------------------------------------------------
 
     template <typename T>
-    inline const std::vector<Directory<T>*> &Directory<T>::subdirs() const
+    inline const typename Directory<T>::DirectoryList &Directory<T>::subdirs() const
     {
       return m_subdirs;
     }
 
     //-------------------------------------------------------------------------------------------------
 
     template <typename T>
-    inline Directory<T> *Directory<T>::mkdir(const std::string &dirName)
+    inline typename Directory<T>::DirectoryPtr Directory<T>::mkdir(const std::string &dirName)
     {
-      Directory<T> *pDirectory = nullptr;
+      DirectoryPtr directory;
```

Comment:
Looking at [http://en.cppreference.com/w/cpp/memory/enable_shared_from_this](http://en.cppreference.com/w/cpp/memory/enable_shared_from_this)
Shouldn't the directory be made with `DirectoryPtr directory = std::make_shared<DirectoryPtr>();` to avoid having some `bad_weak_ptr` dangling around when calling `directory.shared_from_this()`

Reply:
No because in this case calling make_shared will allocate a new directory which I don't want. I'm looking for an already exiting directory to return.

Reply:
Yes, it is quite obvious now that you mention it

Reply:
Yes, it is quite obvious now that you mention it